### PR TITLE
Update tf.cross, preserve quaternion dtype, and new quaternion computing function

### PR DESCRIPTION
**Pull request by ecpoppenheimer**
I made two bug fixes:
1) When the quaternion class needs to create a new quaternion (say in inverse() ) the new quaternion will take the default dtype, instead of the parent's dtype, which will cause a type error if the parent does not use the default dtype.  I added code to pass the parent's dtype to the child in these cases.
2) tf.cross seems to not be a valid alias for tf.linalg.cross.

On the version of this library I obtained from pip the keep_dims / keepdims change was not present, but it looks like you did make this change in the repository.  Perhaps the version pip knows about isn't updated?  (I have no idea how pip works)

I also added a utility function which I find to be the most convenient way of defining a quaternion when rotating a collection of vectors in the same way (i.e. rotating a mesh in CAD without distorting it).  You can ignore this if you aren't interested.

### CODE REVIEW

File: tfquaternion/tfquaternion.py [hunk 1]
```diff
@@ -163,8 +163,42 @@ def rotate_vector_by_quaternion(q, v, q_ndims=None, v_ndims=None):
...
-    t = 2 * tf.cross(q_xyz, v)
-    return v + tf.expand_dims(w, axis=-1) * t + tf.cross(q_xyz, t)
+    t = 2 * tf.linalg.cross(q_xyz, v)
+    return v + tf.expand_dims(w, axis=-1) * t + tf.linalg.cross(q_xyz, t)
+
+
+def quaternion_between_3d(u, v, epsilion=1e-6, dtype=tf.float32):
+    """
+    This function will help define a rotation when you know how you wish to map one specific vector
+    in a collection onto another, and apply that transformation to the whole collection.
```
Comment (PhilJd) [inline tfquaternion/tfquaternion.py]: This function doesn't do anything specific to a collection, right? Could you rephrase this in terms of what this function does, and not what the potential use case might be?

**Review by PhilJd — COMMENTED** [review]
Thanks a lot for the contribution and sorry for the late reply!
I've added a couple of comments.
Could you also change the title and commit messages to describe which bugs you fix and what new functionality you add instead of the generic 'Minor bug fixes + new code'?
Thanks!

**Comment by PhilJd** [comment]
Argh github. Seems like it lost most of my comments. Will redo.

File: tfquaternion/tfquaternion.py [hunk 2]
```diff
@@ -163,8 +163,42 @@ def rotate_vector_by_quaternion(q, v, q_ndims=None, v_ndims=None):
         v = tf.expand_dims(v, axis=0) + tf.zeros_like(q_xyz)
     q_xyz += tf.zeros_like(v)
     v += tf.zeros_like(q_xyz)
-    t = 2 * tf.cross(q_xyz, v)
-    return v + tf.expand_dims(w, axis=-1) * t + tf.cross(q_xyz, t)
+    t = 2 * tf.linalg.cross(q_xyz, v)
+    return v + tf.expand_dims(w, axis=-1) * t + tf.linalg.cross(q_xyz, t)
+
+
```
Comment (PhilJd) [inline tfquaternion/tfquaternion.py:156]: Add @scope_wrapper to improve graph visualization.

File: tfquaternion/tfquaternion.py [hunk 3]
```diff
@@ -163,8 +163,42 @@ def rotate_vector_by_quaternion(q, v, q_ndims=None, v_ndims=None):
...
+def quaternion_between_3d(u, v, epsilion=1e-6, dtype=tf.float32):
+    """
```
Comment (PhilJd) [inline tfquaternion/tfquaternion.py:159]: Please add a one line description here.

File: tfquaternion/tfquaternion.py [hunk 4]
```diff
@@ -163,8 +163,42 @@ def rotate_vector_by_quaternion(q, v, q_ndims=None, v_ndims=None):
         v = tf.expand_dims(v, axis=0) + tf.zeros_like(q_xyz)
     q_xyz += tf.zeros_like(v)
     v += tf.zeros_like(q_xyz)
-    t = 2 * tf.cross(q_xyz, v)
-    return v + tf.expand_dims(w, axis=-1) * t + tf.cross(q_xyz, t)
+    t = 2 * tf.linalg.cross(q_xyz, v)
+    return v + tf.expand_dims(w, axis=-1) * t + tf.linalg.cross(q_xyz, t)
+
+
+def quaternion_between_3d(u, v, epsilion=1e-6, dtype=tf.float32):
```
Comment (PhilJd) [inline tfquaternion/tfquaternion.py]: Could you rename this function to be a bit more descriptive? E.g. `get_rotation_quaternion_from_u_to_v`

Comment (PhilJd) [inline tfquaternion/tfquaternion.py]:
```suggestion
def quaternion_between_3d(u, v, epsilon=1e-6, dtype=tf.float32):
```

File: tfquaternion/tfquaternion.py [hunk 5]
```diff
@@ -163,8 +163,42 @@ def rotate_vector_by_quaternion(q, v, q_ndims=None, v_ndims=None):
...
+    # The dot / cross product determination produces twice the desired rotation.
+    # Half the rotation can be accomplished by averaging the rotation produced here with a zero
+    # rotation, which is accomplished by simply adding 1 to the dot product.
+    dot = tf.expand_dims(tf.tensordot(u, v, 1), 0) + 1
+    cross = tf.linalg.cross(u, v)
+
+    # But if they rotation is a pure 180 flip then this will cause the dot product to become -1
+    # which will create a zero quaternion, which is not what we want.  So have to manually construct
+    # the quaternion in this case
+    if dot < epsilion:
```
Comment (PhilJd) [inline tfquaternion/tfquaternion.py]: Could you use tf.where instead of if here?

File: tfquaternion/tfquaternion.py [hunk 6]
```diff
@@ -163,8 +163,42 @@ def rotate_vector_by_quaternion(q, v, q_ndims=None, v_ndims=None):
...
+	    # need any vector orthogonal to the input.  Try crossing with the x-axis and if that does not
+        # work than crossing with the y axis is guarenteed to...
+        ortho = tf.linalg.cross(u, (1, 0, 0))
+        if tf.reduce_sum(ortho) < epsilion:
```
Comment (PhilJd) [inline tfquaternion/tfquaternion.py]: Same here, please use tf.where instead of if

File: tfquaternion/tfquaternion.py [hunk 7]
```diff
@@ -163,8 +163,42 @@ def rotate_vector_by_quaternion(q, v, q_ndims=None, v_ndims=None):
...
+    # rotation, which is accomplished by simply adding 1 to the dot product.
+    dot = tf.expand_dims(tf.tensordot(u, v, 1), 0) + 1
+    cross = tf.linalg.cross(u, v)
+
+    # But if they rotation is a pure 180 flip then this will cause the dot product to become -1
+    # which will create a zero quaternion, which is not what we want.  So have to manually construct
+    # the quaternion in this case
+    if dot < epsilion:
+	    # need any vector orthogonal to the input.  Try crossing with the x-axis and if that does not
+        # work than crossing with the y axis is guarenteed to...
```
Comment (PhilJd) [inline tfquaternion/tfquaternion.py]: Could you replace the '...' with an explicit description of what is guaranteed?

File: tfquaternion/tfquaternion.py [hunk 8]
```diff
@@ -163,8 +163,42 @@ def rotate_vector_by_quaternion(q, v, q_ndims=None, v_ndims=None):
...
+    v = tf.math.l2_normalize(v)
+
+    # The dot / cross product determination produces twice the desired rotation.
+    # Half the rotation can be accomplished by averaging the rotation produced here with a zero
+    # rotation, which is accomplished by simply adding 1 to the dot product.
+    dot = tf.expand_dims(tf.tensordot(u, v, 1), 0) + 1
+    cross = tf.linalg.cross(u, v)
+
+    # But if they rotation is a pure 180 flip then this will cause the dot product to become -1
+    # which will create a zero quaternion, which is not what we want.  So have to manually construct
```
Comment (PhilJd) [inline tfquaternion/tfquaternion.py]:
```suggestion
    # which will create a zero quaternion, which is not what we want. So have to manually construct
```

File: tfquaternion/tfquaternion.py [hunk 9]
```diff
@@ -163,8 +163,42 @@ def rotate_vector_by_quaternion(q, v, q_ndims=None, v_ndims=None):
...
+    # the quaternion in this case
```
Comment (PhilJd) [inline tfquaternion/tfquaternion.py]:
```suggestion
    # the quaternion in this case.
```

File: tfquaternion/tfquaternion.py [hunk 10]
```diff
@@ -163,8 +163,42 @@ def rotate_vector_by_quaternion(q, v, q_ndims=None, v_ndims=None):
...
+    if dot < epsilion:
+	    # need any vector orthogonal to the input.  Try crossing with the x-axis and if that does not
+        # work than crossing with the y axis is guarenteed to...
+        ortho = tf.linalg.cross(u, (1, 0, 0))
+        if tf.reduce_sum(ortho) < epsilion:
+            ortho = tf.linalg.cross(u, (0, 1, 0))
+        q = tfq.Quaternion(tf.math.l2_normalize(tf.concat([dot, ortho], -1)), dtype=dtype)
+        print(f"did a flip! {q}")
```
Comment (PhilJd) [inline tfquaternion/tfquaternion.py]: Please remove debug prints.

**Review by PhilJd — COMMENTED** [review]
Thanks a lot for the contribution and sorry for the late reply!
I've added a couple of comments.
Could you also change the title and commit messages to describe which bugs you fix and what new functionality you add instead of the generic 'Minor bug fixes + new code'?
Thanks!

**Comment by PhilJd** [comment]
Please also add test cases for the new function!

**Comment by ecpoppenheimer** [comment]
I am still a beginner with git, so probably didn't go about this the correct way.  I updated my fork, will those updates transfer to this pull request?

But the really annoying thing is that I seem to have inadvertently added a lot of changes to whitespace in my most recent commit...  What do you want to do about that?

**Comment by ecpoppenheimer** [comment]
I added in the script that I used to test rotation_from_u_to_v, but it isn't configured correctly as a unit test.  I am a little unsure how to integrate into your unit test class

**Comment by ecpoppenheimer** [comment]
I made some really major changes to u_to_v function, to make it compatible with batches and to fix problems, since it wasn't very well tested when I submitted the PR.  I strongly suspect there is a better way to handle the inversion special case...  We need a quaternion that would take V to -V.  I can't shake the suspicion that there is a really simple way to compute a general case -1 quaternion, but I haven't been able to find one and didn't feel up for trying it by hand.

**Comment by ecpoppenheimer** [comment]
Ah, doesn't work with different float64 objects...

**Comment by ecpoppenheimer** [comment]
Ok, I changed the Quaternion constructor again so that it now takes a default dtype of None, in which case it tries to use the dtype of wxyz, and if that fails then it falls back to tf.float32.

**Comment by ecpoppenheimer** [comment]
So I am really not quite sure what to do with the test cases.  I never learned how to properly use test cases in my limited programming education.  It looks like you are using a built in testing module in TF, is that correct?  Does it automatically call class members with test in their name?  Would it be acceptable to just package all the code in the test script I wrote into the test class in your test script?  I don't want to keep it as it's own file like it is now.

**Comment by ecpoppenheimer** [comment]
I moved the test cases into your existing test class.  Still not sure how that code runs, so I haven't tested the code inside the class, but it does run when it is it's own file.

File: tfquaternion/test_tfquaternion.py [hunk 11]
```diff
@@ -226,6 +226,74 @@ def test_quaternion_to_tensor(self):
...
+            count = u_list.shape[0]
+
+            # Test that we can correctly rotate u onto v individually
+            for i in range(count):
+                u = u_list[i]
+                v = v_list[i]
+                q = tfq.get_rotation_quaternion_from_u_to_v(u, v, epsilon=epsilon)
+                v_computed = tfq.rotate_vector_by_quaternion(q, u)
+
+                assert tf.reduce_all(v - v_computed < epsilon)
```
Comment (PhilJd) [inline tfquaternion/test_tfquaternion.py]: self.assertAllLess

File: tfquaternion/test_tfquaternion.py [hunk 12]
```diff
@@ -226,6 +226,74 @@ def test_quaternion_to_tensor(self):
...
+
+            # Test that flipping u works correctly
+            for i in range(count):
+                u = u_list[i]
+                q = tfq.get_rotation_quaternion_from_u_to_v(u, -u, epsilon=epsilon)
+                neg_u = tfq.rotate_vector_by_quaternion(q, u)
+                assert tf.reduce_all(u + neg_u < epsilon)
```
Comment (PhilJd) [inline tfquaternion/test_tfquaternion.py]: self.assertAllLess

Comment (PhilJd) [reply tfquaternion/test_tfquaternion.py]: (same for the remaining asserts)

File: tfquaternion/test_tfquaternion.py [hunk 13]
```diff
@@ -226,6 +226,74 @@ def test_quaternion_to_tensor(self):
...
+                ),
+                dtype=dtype
+            )
+
+            u_list = tf.concat([u_list, u_manuals], 0)
+            v_list = tf.concat([v_list, v_manuals], 0)
+            count = u_list.shape[0]
+
+            # Test that we can correctly rotate u onto v individually
+            for i in range(count):
```
Comment (PhilJd) [inline tfquaternion/test_tfquaternion.py]: for u, v in zip(u_list, v_list):

File: tfquaternion/test_tfquaternion.py [hunk 14]
```diff
@@ -226,6 +226,74 @@ def test_quaternion_to_tensor(self):
...
+                u = u_list[i]
+                v = v_list[i]
+                q = tfq.get_rotation_quaternion_from_u_to_v(u, v, epsilon=epsilon)
+                v_computed = tfq.rotate_vector_by_quaternion(q, u)
+
+                assert tf.reduce_all(v - v_computed < epsilon)
+
+            # Test that flipping u works correctly
+            for i in range(count):
```
Comment (PhilJd) [inline tfquaternion/test_tfquaternion.py]: for u in u_list:

File: tfquaternion/test_tfquaternion.py [hunk 15]
```diff
@@ -226,6 +226,74 @@ def test_quaternion_to_tensor(self):
             self.assertAllEqual(type(tf.convert_to_tensor(ref)), tf.Tensor)
             # reduce sum internally calls tf.convert_to_tensor
             self.assertAllEqual(tf.reduce_sum(ref), 1.0)
+            
+    def test_rotation_from_u_to_v(self):
```
Comment (PhilJd) [inline tfquaternion/test_tfquaternion.py]:
```suggestion
    def test_get_rotation_quaternion_from_u_to_v(self):
```

File: tfquaternion/tfquaternion.py [hunk 16]
```diff
@@ -163,8 +150,77 @@ def rotate_vector_by_quaternion(q, v, q_ndims=None, v_ndims=None):
...
+    # We need to construct the quaternion (0, x, y, z) where (x, y, z) is any vector
+    # orthogonal to u.  To get this, try u cross x-axis, and in the rare case where
+    # that is zero because u is already the x-axis, then crossing with the y axis is
+    # guarenteed to work, because there is no vector that can be orthogonal to both
+    x_axis = tf.cast(tf.broadcast_to((1, 0, 0), u.shape), dtype)
+    y_axis = tf.cast(tf.broadcast_to((0, 1, 0), u.shape), dtype)
+    ortho_x = tf.linalg.cross(u, x_axis)
+    ortho_y = tf.linalg.cross(u, y_axis)
+    ortho = tf.where(
+        tf.less(tf.reduce_sum(tf.multiply(ortho_x, ortho_x), -1, keepdims=True), epsilon),
```
Comment (PhilJd) [inline tfquaternion/tfquaternion.py]: Could you factor out the condition? I.e.
is_parallel = tf.reduce_sum(tf.multiply(ortho_x, ortho_x), -1, keepdims=True) < epsilon
ortho = tf.where(is_parallel, ...)

File: tfquaternion/tfquaternion.py [hunk 17]
```diff
@@ -180,34 +236,34 @@ class Quaternion(object):
...
         Returns:
             A Quaternion.
-
         Raises:
             ValueError, if wxyz is a `tf.Tensor` and the tensors dtype differs
                 from the given dtype.
             ValueError, if the last dimension of wxyz is not 4.
             TypeError, if dtype is not a float.
         """
+        try:
```
Comment (PhilJd) [inline tfquaternion/tfquaternion.py]: How about making this explicit instead of the try/except? I.e.,
```
if dtype is None:
   dtype = wxyz.dtype if isinstance(wxyz, (tf.Tensor, tf.Variable. np.ndarray)) else tf.float32
```

File: tfquaternion/tfquaternion.py [hunk 18]
```diff
@@ -163,8 +150,77 @@ def rotate_vector_by_quaternion(q, v, q_ndims=None, v_ndims=None):
...
+        [dot, cross], -1), axis=-1), dtype=dtype)
+    q_ortho = vector3d_to_quaternion(ortho)
+    q = tf.where(
+        tf.less(dot, epsilon),
+        q_ortho,
+        q_untested
+    )
+    return Quaternion(q, dtype=dtype)
+
+    """if dot < epsilon:
```
Comment (PhilJd) [inline tfquaternion/tfquaternion.py]: I guess this can be removed?

**Review by PhilJd — COMMENTED** [review]
Thanks for making these changes!  I've added one more round of comments.
RE testing: you're right that the function is executed if it starts with test (and integrating it into the test class is the right way to go :)). The test should be executed by running `python test_tf_quaternion.py`, please let me know if that's not the case, then I'll try to look into it.

**Comment by PhilJd** [comment]
Bt the way, the white spaces are fine, they're enforcing correct pep8 :)

**Comment by ecpoppenheimer** [comment]
I have implemented the most recent round of changes.

When I tried running the test I got a bunch of errors due to running things in eager mode.  Looks like you developed this for the graph execution style of TF.  In eager mode tensors no longer have an .eval() method.

It seems like changing the line return self._q.eval(session=session) to
return np.array(self) works (have to then include numpy at the top of the page).  But then the testing script throws a ton of errors because shapes are getting messed up all over the place.

So I am really not sure what to do about this.  The code works in the limited environment where I need it.  Are you able to successfully run the tests?

**Comment by ecpoppenheimer** [comment]
I have thought a little be more about the special case code in get_rotation_quaternion_from_u_to_v.  I need any vector orthogonal to U.  U cross x-axis will be either orthogonal or zero if U is parallel to the x-axis.  I am detecting this condition as a second special case and then trying U cross Y-axis, which should never fail because U cannot be parallel to both.  But maybe it would simplify the code slightly if instead I computed orthogonal = u cross x-axis + u cross y-axis.  That should always give a non-zero orthogonal vector, right?  It won't be normalized, would have to do that.  But that would eliminate the second special case... would only need a single tf.where statement.  But now that we are always computing two cross products and need to normalize, it would be slightly less performant.

**Comment by PhilJd** [comment]
I'm really sorry about the delay, the last two weeks have been extremely busy.
I haven't tried to run these tests in a while and I'm not at my private laptop at the moment but I'll try to take a look later today.

Re tf.where: I think both versions are fine, feel free to go with the one you prefer. The performance should be the same as even for tf.where both branches are evaluated.

**Comment by ecpoppenheimer** [comment]
I am pretty sure the code itself works just fine.  I have been using it for a while now and have not had any issues.  The unit testing however is totally borked.  I spent a little bit of time today trying to fix them and it is just a mess.  The issue is that you wrote this library for graph mode, which has since largely been depreciated.  So it seems like a lot will have to change with the unit test class.  TF isn't really supporting graph mode compatibility very much.  In the last few days I have dealt with a handful of silly issues where core TF functionality just isn't supported moving across the eager / graph mode boundary.  Like, you can't actually return a ragged tensor out of a TF function back into eager mode.  It just isn't supported, but doesn't crash either.  It just gives back a blank tensor.  But... once I wrote the three or four lines of code to manually implement a ragged tensor it works just fine.

So... I am recommending you just drop testing in graph mode altogether.  If it works in eager mode then the actual quaternion code should just work in graph mode too, I just don't understand the unit tester well enough to get it to test both cases.
